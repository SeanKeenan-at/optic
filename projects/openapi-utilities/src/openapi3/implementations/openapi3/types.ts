import { RulesetDefinition as SpectralRulesetDefinition } from '@stoplight/spectral-core';
import { IChange, IFact } from '../../sdk/types';
import { OpenAPIV3 } from 'openapi-types';
import { Result } from '../../../types';

export type SpectralInput = Extract<
  SpectralRulesetDefinition,
  { extends: any; rules: any }
>['rules'];

export type RuleRunner = {
  runSpectralRules?: (inputs: {
    ruleset: SpectralInput;
    nextFacts: IFact[];
    nextJsonLike: OpenAPIV3.Document;
  }) => Promise<Result[]>;

  runRulesWithFacts: (inputs: {
    context: any;
    nextFacts: IFact[];
    currentFacts: IFact[];
    changelog: IChange[];
    nextJsonLike: OpenAPIV3.Document;
    currentJsonLike: OpenAPIV3.Document;
  }) => Result[];
};
