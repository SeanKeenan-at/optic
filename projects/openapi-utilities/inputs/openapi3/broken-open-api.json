{
  "openapi": "3.0.1",
  "paths": {
    "/api/users/{userId}": {
      "get": {
        "responses": {
          "200": {
            "description": "hello",
            "headers": {
              "isgood": {
                "schema": {}
              }
            },
            "content": {
              "application/json": {
                "schema": {
                  "type": "object",
                  "properties": {
                    "hello": {
                      "type": "array",
                      "items": "no"
                    }
                  }
                }
              }
            }
          },
          "400": {
            "description": "hello",
            "headers": {
              "isgood": {
                "schema": {}
              }
            },
            "content": {
              "application/json": {
                "schema": {
                  "type": "not-real"
                }
              }
            }
          }
        }
      }
    }
  },
  "info": {
    "version": "0.0.0",
    "title": "Empty"
  }
}
