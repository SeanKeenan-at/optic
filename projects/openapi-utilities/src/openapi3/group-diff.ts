import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { OpenAPIV3 } from 'openapi-types';
import { ObjectDiff, typeofDiff } from '../diff/diff';
import {
  constructFactTree,
  getFactForJsonPath,
} from './json-path-interpreters';
import { getLocation, OpenApiV3Traverser } from './traverser';
import { OpenApiV3TraverserFact } from './types';

const SEPARATOR = '-~_~-';

function getEndpointId({
  pathPattern,
  method,
}: {
  pathPattern: string;
  method: string;
}) {
  return `${pathPattern}${SEPARATOR}${method.toUpperCase()}`;
}

function constructTree(spec: OpenAPIV3.Document) {
  const traverser = new OpenApiV3Traverser();
  traverser.traverse(spec);

  return constructFactTree([...traverser.facts()]);
}

export type Diff = {
  trail: string; // The relative path from the diff to the significant node
  change: 'added' | 'changed' | 'removed';
} & (
  | {
      before?: undefined;
      after: string;
    }
  | {
      before: string;
      after: string;
    }
  | {
      before: string;
      after?: undefined;
    }
);

export type Body = {
  diffs: Record<string, Diff[]>;
};

export type Response = {
  diffs: Diff[];
  headers: Diff[];
  contents: Record<string, Body>;
};

export type Endpoint = {
  method: string;
  path: string;
  diffs: Diff[];
  queryParameters: Diff[];
  pathParameters: Diff[];
  cookieParameters: Diff[];
  headerParameters: Diff[];
  request: {
    diffs: Diff[];
    contents: Record<string, Body>;
  };
  responses: Record<string, Response>;
};

export class GroupedDiffs {
  public specification: Diff[];
  public endpoints: Record<string, Endpoint>;
  constructor() {
    this.endpoints = {};
    this.specification = [];
  }

  getOrSetEndpoint(endpointId: string): Endpoint {
    if (this.endpoints[endpointId]) {
      return this.endpoints[endpointId];
    } else {
      const [path, method] = endpointId.split(SEPARATOR);
      const endpoint: Endpoint = {
        method,
        path,
        diffs: [],
        queryParameters: [],
        pathParameters: [],
        cookieParameters: [],
        headerParameters: [],
        request: {
          diffs: [],
          contents: {},
        },
        responses: {},
      };
      this.endpoints[endpointId] = endpoint;
      return endpoint;
    }
  }

  getOrSetRequestBody(endpointId: string, contentType: string): Body {
    const endpoint = this.getOrSetEndpoint(endpointId);
    if (endpoint.request.contents[contentType]) {
      return endpoint.request.contents[contentType];
    } else {
      const requestBody: Body = {
        diffs: {},
      };
      endpoint.request.contents[contentType] = requestBody;
      return requestBody;
    }
  }

  getOrSetResponse(endpointId: string, statusCode: string): Response {
    const endpoint = this.getOrSetEndpoint(endpointId);

    if (endpoint.responses[statusCode]) {
      return endpoint.responses[statusCode];
    } else {
      const response: Response = {
        diffs: [],
        headers: [],
        contents: {},
      };
      endpoint.responses[statusCode] = response;
      return response;
    }
  }

  getOrSetResponseBody(
    endpointId: string,
    statusCode: string,
    contentType: string
  ): Body {
    const response = this.getOrSetResponse(endpointId, statusCode);
    if (response.contents[contentType]) {
      return response.contents[contentType];
    } else {
      const responseBody: Body = {
        diffs: {},
      };
      response.contents[contentType] = responseBody;
      return responseBody;
    }
  }
}

function normalizeRequiredDiff(
  spec: OpenAPIV3.Document,
  fact: OpenApiV3TraverserFact<'body'> | OpenApiV3TraverserFact<'field'>,
  pointers: {
    absolute: string;
    trail: string;
  }
): string[] | null {
  const trailParts = jsonPointerHelpers.decode(pointers.trail);
  const maybeBaseRequiredPath =
    fact.type === 'body' && trailParts[1] === 'required'
      ? jsonPointerHelpers.append(fact.location.jsonPath, 'schema')
      : fact.type === 'field' && trailParts[0] === 'required'
      ? fact.location.jsonPath
      : null;
  const location = getLocation(fact);

  if (!maybeBaseRequiredPath) {
    return null;
  }

  const expectAnArray =
    (fact.type === 'body' && trailParts.length === 2) ||
    (fact.type === 'field' && trailParts.length === 1);
  const expectAString =
    (fact.type === 'body' && trailParts.length === 3) ||
    (fact.type === 'field' && trailParts.length === 2);

  // fetch the required keys
  const raw = jsonPointerHelpers.get(spec, pointers.absolute);

  // Here we'll need to check if the required added is valid and maybe fan out keys to add
  if (
    (Array.isArray(raw) && expectAnArray) ||
    (typeof raw === 'string' && expectAString)
  ) {
    const keysToTest = Array.isArray(raw)
      ? raw.filter((k) => typeof k === 'string')
      : [raw];

    return keysToTest
      .map((k) =>
        jsonPointerHelpers.append(maybeBaseRequiredPath, 'properties', k)
      )
      .filter((p) => jsonPointerHelpers.tryGet(spec, p).match)
      .map((p) =>
        jsonPointerHelpers.relative(
          p,
          fact.location.jsonPath.replace(
            'trail' in location
              ? jsonPointerHelpers.compile(location.trail)
              : '',
            ''
          )
        )
      );
  } else {
    return [];
  }
}

export function groupDiffsByEndpoint(
  specs: { from: OpenAPIV3.Document; to: OpenAPIV3.Document },
  diffs: ObjectDiff[]
) {
  const fromTree = constructTree(specs.from);
  const toTree = constructTree(specs.to);

  const grouped = new GroupedDiffs();

  for (const diff of diffs) {
    const fact =
      diff.after !== undefined
        ? getFactForJsonPath(diff.after, toTree)
        : getFactForJsonPath(diff.before, fromTree);
    if (fact) {
      const trail = jsonPointerHelpers.relative(
        diff.after ?? diff.before,
        fact.location.jsonPath
      );
      const diffToAdd = { ...diff, trail, change: typeofDiff(diff) };
      if (fact.type === 'specification') {
        grouped.specification.push(diffToAdd);
      } else if (fact.type === 'path') {
        // We have a path fact, but we don't want to have to keep looking up each diff, so we'll "convert" the raw diff
        // and just emit endpoint diffs
        const { pathPattern } = getLocation(fact);
        if (diff.before !== undefined && diff.after === undefined) {
          const rawPathObject = jsonPointerHelpers.get(specs.from, diff.before);
          for (const method of Object.keys(rawPathObject)) {
            if (Object.values(OpenAPIV3.HttpMethods).includes(method as any)) {
              const newDiff: Diff = {
                before: jsonPointerHelpers.append(diff.before, method),
                trail: '',
                change: 'removed',
              };
              const endpointId = getEndpointId({ pathPattern, method });
              const endpoint = grouped.getOrSetEndpoint(endpointId);
              endpoint.diffs.push(newDiff);
            }
          }
        } else if (diff.before === undefined && diff.after !== undefined) {
          const rawPathObject = jsonPointerHelpers.get(specs.to, diff.after);
          for (const method of Object.keys(rawPathObject)) {
            if (Object.values(OpenAPIV3.HttpMethods).includes(method as any)) {
              const newDiff: Diff = {
                after: jsonPointerHelpers.append(diff.after, method),
                trail: '',
                change: 'added',
              };
              const endpointId = getEndpointId({ pathPattern, method });
              const endpoint = grouped.getOrSetEndpoint(endpointId);
              endpoint.diffs.push(newDiff);
            }
          }
        }
      } else if (fact.type === 'operation') {
        const { pathPattern, method } = getLocation(fact);
        const endpointId = getEndpointId({ pathPattern, method });
        const endpoint = grouped.getOrSetEndpoint(endpointId);
        endpoint.diffs.push(diffToAdd);
      } else if (fact.type === 'request-header') {
        const { pathPattern, method } = getLocation(fact);
        const endpointId = getEndpointId({ pathPattern, method });
        const endpoint = grouped.getOrSetEndpoint(endpointId);
        endpoint.headerParameters.push(diffToAdd);
      } else if (fact.type === 'request-query') {
        const { pathPattern, method } = getLocation(fact);
        const endpointId = getEndpointId({ pathPattern, method });
        const endpoint = grouped.getOrSetEndpoint(endpointId);
        endpoint.queryParameters.push(diffToAdd);
      } else if (fact.type === 'request-cookie') {
        const { pathPattern, method } = getLocation(fact);
        const endpointId = getEndpointId({ pathPattern, method });
        const endpoint = grouped.getOrSetEndpoint(endpointId);
        endpoint.cookieParameters.push(diffToAdd);
      } else if (fact.type === 'request-path') {
        const { pathPattern, method } = getLocation(fact);
        const endpointId = getEndpointId({ pathPattern, method });
        const endpoint = grouped.getOrSetEndpoint(endpointId);
        endpoint.pathParameters.push(diffToAdd);
      } else if (fact.type === 'requestBody') {
        const { pathPattern, method } = getLocation(fact);
        const endpointId = getEndpointId({ pathPattern, method });
        const endpoint = grouped.getOrSetEndpoint(endpointId);
        endpoint.request.diffs.push(diffToAdd);
      } else if (fact.type === 'response') {
        const { pathPattern, method, statusCode } = getLocation(fact);
        const endpointId = getEndpointId({ pathPattern, method });
        const response = grouped.getOrSetResponse(endpointId, statusCode);
        response.diffs.push(diffToAdd);
      } else if (fact.type === 'response-header') {
        const { pathPattern, method, statusCode } = getLocation(fact);
        const endpointId = getEndpointId({ pathPattern, method });
        const response = grouped.getOrSetResponse(endpointId, statusCode);
        response.headers.push(diffToAdd);
      } else if (fact.type === 'body-example') {
        const location = getLocation(fact);

        const endpointId = getEndpointId(location);
        const key = jsonPointerHelpers.compile(location.trail);
        const body =
          location.location === 'request'
            ? grouped.getOrSetRequestBody(endpointId, location.contentType)
            : grouped.getOrSetResponseBody(
                endpointId,
                location.statusCode,
                location.contentType
              );

        if (body.diffs[key]) {
          body.diffs[key].push(diffToAdd);
        } else {
          body.diffs[key] = [diffToAdd];
        }
      } else if (fact.type === 'body' || fact.type === 'field') {
        const location = getLocation(fact);
        const endpointId = getEndpointId(location);
        const body =
          location.location === 'request'
            ? grouped.getOrSetRequestBody(endpointId, location.contentType)
            : grouped.getOrSetResponseBody(
                endpointId,
                location.statusCode,
                location.contentType
              );

        const specToFetchFrom =
          diff.after !== undefined ? specs.to : specs.from;

        let fieldKeys = normalizeRequiredDiff(specToFetchFrom, fact, {
          absolute: diff.after ?? diff.before,
          trail,
        }) ?? [
          'trail' in location ? jsonPointerHelpers.compile(location.trail) : '',
        ];

        for (const fieldKey of fieldKeys) {
          if (body.diffs[fieldKey]) {
            body.diffs[fieldKey].push(diffToAdd);
          } else {
            body.diffs[fieldKey] = [diffToAdd];
          }
        }
      }
    }
  }

  return grouped;
}
