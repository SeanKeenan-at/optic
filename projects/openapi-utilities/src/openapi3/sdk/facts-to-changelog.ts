import { IFact, IChange, OpenApiFact } from "./types";

import equals from "fast-deep-equal";

export function factsToChangelog(
  past: IFact<OpenApiFact>[],
  current: IFact<OpenApiFact>[]
): IChange<OpenApiFact>[] {
  const added = current.filter(
    (i) =>
      !past.some((fact) =>
        equals(fact.location.conceptualPath, i.location.conceptualPath)
      )
  );
  const removed = past.filter(
    (i) =>
      !current.some((fact) =>
        equals(fact.location.conceptualPath, i.location.conceptualPath)
      )
  );
  const updated = past.filter((i) => {
    const currentVersion = current.find((fact) =>
      equals(fact.location.conceptualPath, i.location.conceptualPath)
    );
    if (currentVersion) {
      if (equals(i.value, currentVersion.value)) {
        return false;
      } else return true;
    } else return false;
  });

  return [
    ...added.map((added) => ({
      location: added.location,
      added: added.value,
    })),
    ...removed.map((removed) => ({
      location: removed.location,
      removed: {
        before: removed.value,
      },
    })),
    ...updated.map((past) => {
      const after = current.find((fact) =>
        equals(fact.location.conceptualPath, past.location.conceptualPath)
      )!;
      return {
        location: past.location,
        changed: {
          before: past.value,
          after: after.value,
        },
      };
    }),
  ];
}
