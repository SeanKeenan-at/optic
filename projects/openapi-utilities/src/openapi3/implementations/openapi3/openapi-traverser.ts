import { FactAccumulator, Traverse } from "../../sdk/types";
import { IPathComponent } from "../../sdk/types";
import invariant from "ts-invariant";
import { OpenAPIV3 } from "@useoptic/common";
import { jsonPointerHelpers as jsonPointer } from "@useoptic/common";

export function normalizeOpenApiPath(path: string): string {
  return path
    .split("/")
    .map((pathComponent) => {
      if (pathComponent.startsWith("{") && pathComponent.endsWith("}")) {
        return "{}";
      }
      return pathComponent;
    })
    .join("/");
}

export class OpenAPITraverser
  implements Traverse<OpenAPIV3.Document, OpenApiFact>
{
  format = "openapi3";
  accumulator = new FactAccumulator<OpenApiFact>([]);

  input: OpenAPIV3.Document | undefined = undefined;

  traverse(input: OpenAPIV3.Document): void {
    this.input = input;
    Object.entries(input.paths).forEach(([pathPattern, paths]) => {
      if (paths?.get)
        this.traverseOperations(paths?.get!, "get", pathPattern, {
          method: "get",
          path: pathPattern,
        });
      if (paths?.patch)
        this.traverseOperations(paths?.patch!, "patch", pathPattern, {
          method: "patch",
          path: pathPattern,
        });
      if (paths?.post)
        this.traverseOperations(paths?.post!, "post", pathPattern, {
          method: "post",
          path: pathPattern,
        });
      if (paths?.put)
        this.traverseOperations(paths?.put!, "put", pathPattern, {
          method: "put",
          path: pathPattern,
        });
      if (paths?.delete)
        this.traverseOperations(paths?.delete!, "delete", pathPattern, {
          method: "delete",
          path: pathPattern,
        });
    });
  }

  traverseOperations(
    operation: OpenAPIV3.OperationObject,
    method: string,
    pathPattern: string,
    location: ConceptualLocation
  ): void {
    const jsonPath = jsonPointer.append("", "paths", pathPattern, method);
    this.checkJsonTrail(jsonPath, operation);
    const normalizedPath = normalizeOpenApiPath(pathPattern);
    const conceptualPath = ["operations", normalizedPath, method];
    this.onOperation(
      operation,
      pathPattern,
      method,
      jsonPath,
      conceptualPath,
      location
    );

    this.traverseParameters(
      operation,
      jsonPointer.append(jsonPath, "parameters"),
      [...conceptualPath, "parameters"],
      location
    );

    if (operation.requestBody) {
      const requestBody = operation.requestBody as OpenAPIV3.RequestBodyObject;
      Object.entries(requestBody.content || {}).forEach(
        ([contentType, body]) => {
          this.traverseBody(
            body,
            contentType,
            jsonPointer.append(jsonPath, "requestBody", "content", contentType),
            [...conceptualPath, contentType],
            { ...location, inRequest: { body: { contentType } } }
          );
        }
      );
    }

    Object.entries(operation.responses).forEach(([statusCode, response]) => {
      this.traverseResponse(
        response as OpenAPIV3.ResponseObject,
        statusCode,
        jsonPointer.append(jsonPath, "responses", statusCode),
        [...conceptualPath, "responses", statusCode],
        { ...location, inResponse: { statusCode } }
      );
    });
  }

  traverseResponse(
    response: OpenAPIV3.ResponseObject,
    statusCode: string,
    jsonPath: string,
    conceptualPath: IPathComponent[],
    location: ConceptualLocation
  ): void {
    this.traverseResponseHeaders(response, jsonPath, conceptualPath, location);
    this.checkJsonTrail(jsonPath, response);

    Object.entries(response.content || {}).forEach(([contentType, body]) => {
      this.traverseBody(
        body,
        contentType,
        jsonPointer.append(jsonPath, "content", contentType),
        [...conceptualPath, contentType],
        {
          ...location,
          inResponse: {
            statusCode: location.inResponse!.statusCode,
            body: { contentType },
          },
        }
      );
    });

    this.onResponse(response, statusCode, jsonPath, conceptualPath, location);
  }

  traverseParameters(
    operation: OpenAPIV3.OperationObject,
    jsonPath: string,
    conceptualPath: IPathComponent[],
    location: ConceptualLocation
  ) {
    if (operation.parameters) {
      this.checkJsonTrail(jsonPath, operation.parameters);
      operation.parameters.forEach((p, i) => {
        const parameter = p as OpenAPIV3.ParameterObject;

        const locationForParameter = (() => {
          if (parameter.in === "query")
            return { ...location, inRequest: { query: parameter.name } };
          if (parameter.in === "header")
            return { ...location, inRequest: { header: parameter.name } };
          if (parameter.in === "path")
            return { ...location, inRequest: { path: parameter.name } };
          // @todo add cookie
          return location;
        })();

        this.onRequestParameter(
          parameter,
          jsonPointer.append(jsonPath, i.toString()),
          [...conceptualPath, parameter.in, parameter.name],
          locationForParameter
        );
      });
    }
  }
  onRequestParameter(
    parameter: OpenAPIV3.ParameterObject,
    jsonPath: string,
    conceptualPath: IPathComponent[],
    location: ConceptualLocation
  ) {
    this.checkJsonTrail(jsonPath, parameter);
    const value: OpenApiRequestParameterFact = {
      ...parameter,
    };

    const inToType = (() => {
      switch (parameter.in) {
        case "query":
          return OpenApiKind.QueryParameter;
        case "header":
          return OpenApiKind.HeaderParameter;
        case "path":
          return OpenApiKind.PathParameter;
        default:
          return;
      }
    })();

    if (inToType) {
      this.accumulator.log({
        location: {
          jsonPath,
          conceptualPath,
          kind: inToType,
          conceptualLocation: location,
        },
        value,
      });
    }
  }

  traverseResponseHeaders(
    response: OpenAPIV3.ResponseObject,
    jsonPath: string,
    conceptualPath: IPathComponent[],
    location: ConceptualLocation
  ) {
    if (response.headers) {
      Object.entries(response.headers).forEach(([name, value]) => {
        const header = value as OpenAPIV3.HeaderObject;
        this.onResponseHeader(
          name,
          header,
          jsonPointer.append(jsonPath, "headers", name),
          [...conceptualPath, "headers", name],
          {
            ...location,
            inResponse: {
              statusCode: location.inResponse!.statusCode,
              header: name,
            },
          }
        );
      });
    }
  }

  onResponseHeader(
    name: string,
    header: OpenAPIV3.HeaderObject,
    jsonPath: string,
    conceptualPath: IPathComponent[],
    location: ConceptualLocation
  ) {
    this.checkJsonTrail(jsonPath, header);
    const value: OpenApiHeaderFact = {
      name,
      ...header,
    };
    this.accumulator.log({
      location: {
        jsonPath,
        conceptualPath,
        kind: OpenApiKind.ResponseHeader,
        conceptualLocation: location,
      },
      value,
    });
  }

  traverseLinks() {}

  traverseBody(
    body: OpenAPIV3.MediaTypeObject,
    contentType: string,
    jsonPath: string,
    conceptualPath: IPathComponent[],
    location: ConceptualLocation
  ) {
    this.checkJsonTrail(jsonPath, body);
    //@TODO: not sure if we need to check the body.schema key count
    if (body.schema && Object.keys(body.schema).length) {
      this.onContentForBody(
        body,
        contentType,
        jsonPath,
        conceptualPath,
        location
      );
      this.traverseSchema(
        body.schema as OpenAPIV3.SchemaObject,
        jsonPointer.append(jsonPath, "schema"),
        conceptualPath,
        location
      );
    }
  }

  traverseField(
    key: string,
    schema: OpenAPIV3.SchemaObject,
    required: boolean,
    jsonPath: string,
    conceptualPath: IPathComponent[],
    location: ConceptualLocation
  ) {
    this.checkJsonTrail(jsonPath, schema);
    this.onField(key, schema, required, jsonPath, conceptualPath, location);
    this.traverseSchema(schema, jsonPath, conceptualPath, location);
  }

  traverseSchema(
    schema: OpenAPIV3.SchemaObject,
    jsonPath: string,
    conceptualPath: IPathComponent[],
    location: ConceptualLocation
  ) {
    this.checkJsonTrail(jsonPath, schema);
    if (schema.oneOf || schema.anyOf || schema.allOf) {
      // iterate these, multiple branches at path
    }
    switch (schema.type) {
      case "object":
        // this.onObject(...)
        Object.entries(schema.properties || {}).forEach(([key, fieldSchema]) =>
          this.traverseField(
            key,
            fieldSchema as OpenAPIV3.SchemaObject,
            (schema.required || []).includes(key),
            jsonPointer.append(jsonPath, "properties", key),
            [...conceptualPath, key],
            {
              ...location,
              jsonSchemaTrail: [...(location.jsonSchemaTrail || []), key],
            }
          )
        );
        break;
      case "array":
        // this.onArray()
        this.traverseSchema(
          schema.items as OpenAPIV3.SchemaObject,
          jsonPointer.append(jsonPath, "items"),
          [...conceptualPath, "items"],
          {
            ...location,
            jsonSchemaTrail: [...(location.jsonSchemaTrail || []), "items"],
          }
        );
        break;
      case "string":
      case "number":
      case "integer":
        break;
    }
  }

  ///////////////////////////////////////////////////////////////////////////////////

  getSchemaWithoutNestedThings(schema: OpenAPIV3.SchemaObject) {
    //@ts-ignore
    const { items, required, properties, ...schemaWithoutNestedThings } =
      schema;
    return schemaWithoutNestedThings as OpenAPIV3.SchemaObject;
  }

  onContentForBody(
    body: OpenAPIV3.MediaTypeObject,
    contentType: string,
    jsonPath: string,
    conceptualPath: IPathComponent[],
    location: ConceptualLocation
  ) {
    const schema = body.schema! as OpenAPIV3.SchemaObject;
    const flatSchema = this.getSchemaWithoutNestedThings(schema);
    const value: OpenApiBodyFact = {
      contentType,
      flatSchema,
    };
    this.accumulator.log({
      location: {
        jsonPath,
        conceptualPath,
        kind: OpenApiKind.Body,
        conceptualLocation: location,
      },
      value,
    });
  }

  onField(
    key: string,
    schema: OpenAPIV3.SchemaObject,
    required: boolean,
    jsonPath: string,
    conceptualPath: IPathComponent[],
    location: ConceptualLocation
  ) {
    this.checkJsonTrail(jsonPath, schema);
    const flatSchema = this.getSchemaWithoutNestedThings(schema);

    const value: OpenApiFieldFact = {
      key,
      flatSchema,
      required,
    };
    this.accumulator.log({
      location: {
        jsonPath,
        conceptualPath,
        kind: OpenApiKind.Field,
        conceptualLocation: location,
      },
      value,
    });
  }

  getOperationWithoutNestedThings(
    operation: OpenAPIV3.OperationObject,
    location: ConceptualLocation
  ): OpenAPIV3.OperationObject {
    const { parameters, responses, ...operationWithoutNestedThings } =
      operation;
    return operationWithoutNestedThings as OpenAPIV3.OperationObject;
  }

  onOperation(
    operation: OpenAPIV3.OperationObject,
    pathPattern: string,
    method: string,
    jsonPath: string,
    conceptualPath: IPathComponent[],
    location: ConceptualLocation
  ) {
    this.checkJsonTrail(jsonPath, operation);
    const flatOperation = this.getOperationWithoutNestedThings(
      operation,
      location
    );
    const value: OpenApiOperationFact = {
      ...flatOperation,
      method,
      pathPattern,
    };
    this.accumulator.log({
      location: {
        jsonPath,
        conceptualPath,
        kind: OpenApiKind.Operation,
        conceptualLocation: location,
      },
      value,
    });
  }
  onResponse(
    response: OpenAPIV3.ResponseObject,
    statusCode: string,
    jsonPath: string,
    conceptualPath: IPathComponent[],
    location: ConceptualLocation
  ) {
    this.checkJsonTrail(jsonPath, response);
    const flatResponse = this.getResponseWithoutNestedThings(response);
    const value: OpenApiResponseFact = {
      ...flatResponse,
      statusCode: parseInt(statusCode),
    };
    this.accumulator.log({
      location: {
        jsonPath,
        conceptualPath,
        kind: OpenApiKind.Response,
        conceptualLocation: location,
      },
      value,
    });
  }
  getResponseWithoutNestedThings(
    response: OpenAPIV3.ResponseObject
  ): Omit<OpenAPIV3.ResponseObject, "headers" | "content"> {
    const { headers, content, ...responseWithoutNestedThings } = response;
    return responseWithoutNestedThings as OpenAPIV3.ResponseObject;
  }

  // helper

  checkJsonTrail<G>(jsonPath: string, mustShareIdentity: G) {
    invariant(
      jsonPointer.get(this.input!, jsonPath.toString()) === mustShareIdentity,
      `json trail is not being set properly at ${jsonPath}`
    );
  }
}

//@TODO: figure out what other info downstream tools will need
export interface ConceptualLocation {
  path: string;
  method: string;
  inRequest?: {
    header?: string;
    path?: string;
    query?: string;
    body?: {
      contentType: string;
    };
  };
  inResponse?: {
    header?: string;
    query?: string;
    body?: {
      contentType: string;
    };
    statusCode: string;
  };
  jsonSchemaTrail?: string[];
}

export enum OpenApiKind {
  Operation = "operation",
  Request = "request",
  QueryParameter = "query-parameter",
  PathParameter = "path-parameter",
  HeaderParameter = "header-parameter",
  ResponseHeader = "response-header",
  Response = "response",
  Body = "body",
  Object = "object",
  Field = "field",
  Array = "array",
  Primitive = "primitive",
}

type OpenApiFact =
  | OpenApiOperationFact
  | OpenApiRequestFact
  | OpenApiRequestParameterFact
  | OpenApiResponseFact
  | OpenApiHeaderFact
  | OpenApiBodyFact
  | OpenApiFieldFact;

export interface OpenApiOperationFact extends OpenAPIV3.OperationObject {
  pathPattern: string;
  method: string;
}

export interface OpenApiBodyFact {
  contentType: string;
  flatSchema: OpenAPIV3.SchemaObject;
}

export interface OpenApiFieldFact {
  key: string;
  required: boolean;
  flatSchema: OpenAPIV3.SchemaObject;
}
export interface OpenApiResponseFact
  extends ReturnType<OpenAPITraverser["getResponseWithoutNestedThings"]> {
  statusCode: number;
}
export interface OpenApiRequestFact {}

export interface OpenApiHeaderFact extends OpenAPIV3.HeaderObject {
  name: string;
}

export interface OpenApiRequestParameterFact
  extends OpenAPIV3.ParameterObject {}
