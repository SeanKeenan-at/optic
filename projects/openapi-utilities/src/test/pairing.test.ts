import { OpenAPIV3 } from "openapi-types";
import tap from "tap";
import { OpenAPITraverser } from "../openapi3/implementations/openapi3/OpenAPITraverser";
import { factsToChangelog } from "../openapi3/sdk/facts-to-changelog";

async function checkAdvancedSchemaUsage() {
    const fileContents: OpenAPIV3.Document = {
        openapi: '3.0.3',
        info: {
            title: 'MyService',
            version: '3.0.0',
        },
        servers: [],
        components: {
            headers: {},
            parameters: {},
            responses: {},
            schemas: {},
        },
        paths: {},
    };

    const fileContentsWithChanges: OpenAPIV3.Document = {
        ...fileContents,
    };
    fileContentsWithChanges.paths = {
        '/example': {
            get: {
                responses: {
                    "200": {
                        "description": "d",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "object",
                                    "properties": {
                                        "s": {
                                            type: "string"
                                        }
                                    }
                                }
                            }
                        }
                    },
                },
            },
        },
    }
    const fileContentsWithChanges2: OpenAPIV3.Document = {
        ...fileContents,
    };
    fileContentsWithChanges2.paths = {
        '/example': {
            get: {
                responses: {
                    "200": {
                        "description": "d",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "object",
                                    "properties": {
                                        "s": {
                                            type: "boolean"
                                        }
                                    }
                                }
                            }
                        }
                    },
                },
            },
        },
    }

    tap.test("nested requires", async (t) => {
        const traverser = new OpenAPITraverser();
        traverser.traverse(fileContentsWithChanges);
        const facts = traverser.accumulator.allFacts();

        const traverser2 = new OpenAPITraverser();
        traverser2.traverse(fileContentsWithChanges2);
        const facts2 = traverser2.accumulator.allFacts();

        console.log(facts)
        t.matchSnapshot(facts);
        const changes = factsToChangelog(facts, facts2)
        t.matchSnapshot(changes);
        t.end();
    })

    tap.test("diff", async (t) => {

    })
}

async function main() {
    checkAdvancedSchemaUsage()
}

main();

/*
    Things to cover at some point
    - openapi version field
    - info block
    - servers block

    body should just have content type and schema root type
    schema children should have their own nodes


*/