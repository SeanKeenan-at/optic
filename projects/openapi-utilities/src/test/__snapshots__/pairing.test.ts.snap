// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nested requires 1`] = `
Array [
  Object {
    "location": Object {
      "conceptualLocation": Object {
        "method": "get",
        "path": "/example",
      },
      "conceptualPath": Array [
        "operations",
        "/example",
        "get",
      ],
      "jsonPath": "/paths/~1example/get",
      "kind": "operation",
    },
    "value": Object {
      "method": "get",
      "pathPattern": "/example",
    },
  },
  Object {
    "location": Object {
      "conceptualLocation": Object {
        "inResponse": Object {
          "body": Object {
            "contentType": "application/json",
          },
          "statusCode": "200",
        },
        "method": "get",
        "path": "/example",
      },
      "conceptualPath": Array [
        "operations",
        "/example",
        "get",
        "responses",
        "200",
        "application/json",
      ],
      "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/body",
      "kind": "body",
    },
    "value": Object {
      "contentType": "application/json",
      "flatSchema": Object {
        "type": "object",
      },
    },
  },
  Object {
    "location": Object {
      "conceptualLocation": Object {
        "inResponse": Object {
          "body": Object {
            "contentType": "application/json",
          },
          "statusCode": "200",
        },
        "jsonSchemaTrail": Array [
          "s",
        ],
        "method": "get",
        "path": "/example",
      },
      "conceptualPath": Array [
        "operations",
        "/example",
        "get",
        "responses",
        "200",
        "application/json",
        "s",
      ],
      "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/body/properties/s",
      "kind": "field",
    },
    "value": Object {
      "flatSchema": Object {
        "type": "string",
      },
      "key": "s",
      "required": false,
    },
  },
  Object {
    "location": Object {
      "conceptualLocation": Object {
        "inResponse": Object {
          "statusCode": "200",
        },
        "method": "get",
        "path": "/example",
      },
      "conceptualPath": Array [
        "operations",
        "/example",
        "get",
        "responses",
        "200",
      ],
      "jsonPath": "/paths/~1example/get/responses/200",
      "kind": "response",
    },
    "value": Object {
      "description": "d",
      "statusCode": 200,
    },
  },
]
`;

exports[`nested requires 2`] = `
Array [
  Object {
    "changed": Object {
      "after": Object {
        "flatSchema": Object {
          "type": "boolean",
        },
        "key": "s",
        "required": false,
      },
      "before": Object {
        "flatSchema": Object {
          "type": "string",
        },
        "key": "s",
        "required": false,
      },
    },
    "location": Object {
      "conceptualLocation": Object {
        "inResponse": Object {
          "body": Object {
            "contentType": "application/json",
          },
          "statusCode": "200",
        },
        "jsonSchemaTrail": Array [
          "s",
        ],
        "method": "get",
        "path": "/example",
      },
      "conceptualPath": Array [
        "operations",
        "/example",
        "get",
        "responses",
        "200",
        "application/json",
        "s",
      ],
      "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/body/properties/s",
      "kind": "field",
    },
  },
]
`;
